### PR TITLE
Reland "Disable terser on polymer bundle in dev assets mode (#4651)"

### PR DESCRIPTION
**Pull request by psybuzz**
**Status**: this PR was originally a roll-forward, but that was
abandoned after we reverted (#4655) the offending sync blocker.
This PR is now a re-land of the original #4651 with modifications.

Re-lands the change in #4651 that was reverted in #4655.
The original change used the `compile` option on `tf_js_binary`
as a signal to skip the Terser minification step. However, this
broke internally because `compile=False` has special semantics
in Google.

Now, the reland does the same minification skip when a new
rule attribute `dev_mode_only=True` is present. Internally,
this attribute is ignored.

Googlers, see test sync cl/355919112.

Before this is landed, cl/355772570 must be submitted first,
to fix syncing issues.

### CODE REVIEW

**Review by wchargin — COMMENTED** [review]
Hmm; my expectation is generally that tags are not load-bearing by Bazel
rules themselves, so this feels surprising to me. Might we instead pass
`dev_only = True` to the macro itself, and use that signal to control
the behavior? (Might need an internal `kwargs.pop`, but that can be made
without having to land this atomically.)

**Comment by psybuzz** [comment]
Avoiding a load-bearing tag sounds good.  I've replaced it with a "dev_mode_only" attribute and added a `kwargs.pop` internally at cl/355772570.

**Comment by psybuzz** [comment]
Actually, just in case there's anything controversial in this PR, it's probably better to immediately revert and unblock the sync.  If it's more comfortable, I have a revert PR out at #4655

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:58]: I’m confused about a bunch of things, so maybe I’m just misunderstanding
it all. It looks to me like, as written, `dev_mode_only` will be passed
through to `rollup_bundle` as `kwargs`, which is an error. I expected to
see something like:

```python
def tf_js_binary(
    name,
    compile,
    deps,
    visibility = None,
    dev_mode_only = False,
    **kwargs
):
    ...  # now `rollup_bundle(..., **kwargs)` is okay
```

But it also looks to me like the `dev_mode_only` condition here is
backward, and the real `tf_js_binary` target doesn’t actually set
`dev_mode_only = True`, so I’m not convinced that I understand…

In any case, agreed to start with a rollback.

Comment (psybuzz) [reply tensorboard/defs/defs.bzl]: Sorry for the confusion, I made some mistakes and didn't actually upload the patch with the right changes.  You're totally correct that the `tf_js_binary` wasn't setting the flag, and didn't work.

Now that the offending PR was reverted, I've updated this PR to be a re-land of the original.  The original commit is the first one, and the 2nd commit in this branch contains new modifications.

The test sync cl has also been updated: cl/355900608

File: tensorboard/webapp/dev_assets/BUILD
Comment (wchargin) [inline tensorboard/webapp/dev_assets/BUILD:43]: We keep `compile` now; update comment?

(Or, is the comment needed now? This is in the `dev_assets` package, and
sets `dev_mode_only = True`, so is it self-explanatory? I defer to you.)

File: tensorboard/webapp/dev_assets/BUILD
Comment (wchargin) [inline tensorboard/webapp/dev_assets/BUILD:70]: …fwiw, this feels pretty sketch to me; it could have false-positive
matches, and if `index.html` changes structurally then this replacement
could become a silent no-op. But I see that this was approved in #4651,
so I won’t block on it.

It would make me very slightly more comfortable to anchor the pattern to
the `\.js` extension so that it’s at least idempotent.

File: tensorboard/defs/defs.bzl
Comment (wchargin) [inline tensorboard/defs/defs.bzl:63]: nit: Inverting the control flow here (`if dev_mode_only:`) both makes
the shorter block come first and avoids a negation, both of which are
generally considered easier to read.

Other than that, this logic looks good to me.

**Review by wchargin — APPROVED** [review]
LGTM once <http://cl/355772570> is submitted and a test sync passes.

**Review by psybuzz — COMMENTED** [review]
Thanks for the review

File: tensorboard/webapp/dev_assets/BUILD
Comment (wchargin) [reply tensorboard/webapp/dev_assets/BUILD]: > I also a comment in the *.html saying that a substitution may take place.

Good idea; appreciate it.

**Pull request merged**
